### PR TITLE
Enhance UI 

### PR DESCRIPTION
**Pull request by bangxiangyong**
Main purpose is to improve the UI to be more `stand-alone` , which should help in dockerisation later on too.

- Added some clean up to the UI such as removing the "Add Datastream agent" which is not functioning anymore
- Agent developers can specify the dictionary of parameter choices in their agent  design, and these will be dynamically shown in the UI (see example in the develop_UI folder (this folder is to be removed eventually))
- Changed from Process to Thread in starting the Dashboard (https://github.com/Met4FoF/agentMET4FOF/commit/e2991ddc26ac9dea863d25a9643983a3aad0ac8e) . Not sure if there are any side-effect ?

Latest UI update on agent and coalition instantiation:
image

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -58,3 +58,4 @@ tqdm==4.53.0              # via mesa
 uncertainties==3.1.5      # via time-series-buffer
 urllib3==1.26.2           # via requests
 werkzeug==1.0.1           # via flask
+visdcc==0.0.40
```
Comment (BjoernLudwigPTB) [inline requirements.txt:61]: We need to add this dependency to `setup.py` as well.

File: agentMET4FOF/dashboard/LayoutHelper.py [hunk 2]
```diff
@@ -101,3 +102,28 @@ def visualise_agent_parameters(k,v):
...
+    """
+    Extracts parameters from the init_param dropdown list.
+    These extracted parameters will be passed to the agent's initialisation in the add agent button.
+    """
+    init_params = {}
+    for div in params_div:
+        if isinstance(div,dict):
+            if 'value' in div['props'].keys():
+                init_params.update({div['props']['placeholder']:div['props']['value']})
+    return init_params
```
Comment (BjoernLudwigPTB) [inline agentMET4FOF/dashboard/LayoutHelper.py:129]: Add a new line at the end of the file.

```suggestion
    return init_params

```

**Review by BjoernLudwigPTB — COMMENTED** [review]
This looks really promising. Would you mind me adding type hints specifying the parameter and return types of your methods?

File: requirements.txt (hunk 1 above)
Comment (bangxiangyong) [reply requirements.txt:61]: yes, the visdcc is quite a temporary fix to launch some javascript code callbacks (which was also done with the mpld3 without visdcc, but with a bit more implementation). ideally we wont need the visdcc later on, and just use the built-in dash methods to do the same (like what we did with mpld3).

**Comment by bangxiangyong** [comment]
yes please help in the typehints, i had some problems with circular imports everywhere earlier..

the real issue with the (multi)Process is really in the 'pickling' the modules (which have agents in them) to be included in the dashboard. i saw some suggestion that 'dill' is able to serialise the modules which may work in place.  https://stackoverflow.com/questions/19984152/what-can-multiprocessing-and-dill-do-together

ultimately, the test should be something like : https://github.com/Met4FoF/agentMET4FOF/blob/enhance_UI/develop_UI/agent_main_example.py with the crucial line:
`agent_network = AgentNetwork(dashboard_modules=[agent_module_example], backend="mesa")`
in which the `Process` method (due to pickling the module), and `Thread` did not

File: requirements.txt (hunk 1 above)
Comment (BjoernLudwigPTB) [reply requirements.txt:61]: Ah, I see, so anyway we will not release anything with this in the requirements, right? Then we do not need to add it to `setup.py`.

**Comment by BjoernLudwigPTB** [comment]
> yes please help in the typehints, i had some problems with circular imports everywhere earlier..
> 
> the real issue with the (multi)Process is really in the 'pickling' the modules (which have agents in them) to be included in the dashboard. i saw some suggestion that 'dill' is able to serialise the modules which may work in place. https://stackoverflow.com/questions/19984152/what-can-multiprocessing-and-dill-do-together
> 
> ultimately, the test should be something like : https://github.com/Met4FoF/agentMET4FOF/blob/enhance_UI/develop_UI/agent_main_example.py with the crucial line:
> `agent_network = AgentNetwork(dashboard_modules=[agent_module_example], backend="mesa")`
> in which the `Process` method (due to pickling the module) failed, and `Thread` did not

I did not get that yet. For me both backends run in _agent_main_example.py_ (so having commented out line 7 or 8. Maybe we can talk about it tomorrow morning? That be great!

**Comment by bangxiangyong** [comment]
i meant the breaking would happen without the change: https://github.com/Met4FoF/agentMET4FOF/pull/176/commits/e2991ddc26ac9dea863d25a9643983a3aad0ac8e#diff-49c1382cfbe11028382b0cd2257fc2c79cf95e25778aa9503127f142f9e4fadfR1167

**Comment by BjoernLudwigPTB** [comment]
> i meant the breaking would happen without the change: [e2991dd#diff-49c1382cfbe11028382b0cd2257fc2c79cf95e25778aa9503127f142f9e4fadfR1167](https://github.com/Met4FoF/agentMET4FOF/commit/e2991ddc26ac9dea863d25a9643983a3aad0ac8e#diff-49c1382cfbe11028382b0cd2257fc2c79cf95e25778aa9503127f142f9e4fadfR1167)

Ah yes, of course, sorry. If the use of `multiprocess.Process` does not provide any advantage, we could simply drop it then, yes.

**Comment by bangxiangyong** [comment]
there seemed to be some conflicting files @BjoernLudwigPTB , can you help in resolving this ?

**Comment by bangxiangyong** [comment]
I think we should maintain `Process` in launching dashboard for osbrain due to problems which arise when distributed systems are involved.

i've implemented an easy fix for the "unpicklable" modules which is here: 
https://github.com/Met4FoF/agentMET4FOF/pull/187

**Comment by BjoernLudwigPTB** [comment]
> there seemed to be some conflicting files @BjoernLudwigPTB , can you help in resolving this ?

I will look at it tomorrow morning.

**Comment by BjoernLudwigPTB** [comment]
So we keep the `visdcc` for now. Shall we open an issue to describe what we want to do later, to get rid of it, again?

**Pull request merged**